src/java/org/apache/cassandra/db/streaming/CassandraOutgoingFile.java

```diff
@@ -172,10 +174,11 @@ public void write(StreamSession session, StreamingDataOutputPlus out, int versio
             out.flush();
 
             CassandraStreamWriter writer = header.isCompressed() ?
-                                           new CassandraCompressedStreamWriter(sstable, header, session) :
-                                           new CassandraStreamWriter(sstable, header, session);
+                                           new CassandraCompressedStreamWriter(sstable, header, session, fileStreamMetricsListener) :
+                                           new CassandraStreamWriter(sstable, header, session, fileStreamMetricsListener);
             writer.write(out);
         }
+        fileStreamMetricsListener.onStreamSuccessful();
```

Comment:
what if `onStreamSuccessful` returned boolean so it would be called `isStreamSuccessful` and if it is not it would throw an exception? This method already throws `IOException` so it seems to be quite logical to me that if streaming was not successful we would throw as well. Right now we are doing `assert` in it which can throw `AssertionError` but that is not subclass of `IOException`.

Reply:
I don't think `FileStreamMetricsListener` should be responsible for determining whether a stream is successful, since this class is meant to be a dumb consumer. For instance, I think we should update the number of incoming files streamed on this method `metrics.countStreamedIn(isEntireSSTable)`. The assertion `assert lastSeenBytes == totalSize;` on `onStreamSuccessful()` is just to verify that the behavior of the class `FileStreamMetricsListener` is correct, not to check whether a stream is successful or not.

Reply:
I tried making the change to `isStreamSuccessful` from `onStreamSuccessful` and it seems to make the `FileStreamMetricsListener` a very leaky abstraction to provide a proper error message and we end up duplicating the checking logic between the `CassandraIncomingFile` and `CassandraOutgoingFile` which I'm not a huge fan of (and again, seems like a leaky abstraction).

I could instead throw an `IOException` instead of running the assert; but, like you say, that feels strange as no IO is actually happening in `onStreamSuccessful`.

Reply:
Yes I get that but code-wise that method might throw AssertionError but the method that method is called in throws IOException and that is just wrong.

Reply:
> that method might throw AssertionError but the method that method is called in throws IOException and that is (technically) just wrong.

I don't see what is wrong about this, as far as I understand assertions can be added anywhere in the code to validate assumptions, without requiring methods to throw `AssertionError` explicitly. Am I missing something?

Reply:
OK Paulo.

May I ask, what happens when it actually throws that exception (AssertionError). What happens in runtime? Where it propagates and what eventually fails?

Just honestly asking as I have never checked out this code in IDE to dig deeper, sorry ...

Reply:
> May I ask, what happens when it actually throws that exception (AssertionError). What happens in runtime? Where it propagates and what eventually fails?

The way I see it is that assertions shouldn't happen in runtime unless the code is broken, as they're errors targeted for developers to detect broken code assumptions. Unfortunately in Cassandra we overload assertions a bit and use it for error handling (which IMO is a bad smell), but this is not what is being done here.

The assertion `assert lastSeenBytes == totalSize` on `FileStreamMetricsListener.onStreamSuccessful` is a runtime safety check to ensure that the behavior of the class `FileStreamMetricsListener` is working as expected and we don't report more streamed bytes than the expected amount. It's a future-proof safety check for programming errors, but should not be expected to fail during correct operation.  If this assertion is ever thrown, it means that there is a bug in the code that must be fixed.

Reply:
For a sanity check, we can look into whether or not this can happen during valid operation before we commit this assertion. As far as I can tell it can't, but it's worth digging into further.